src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -145,21 +147,31 @@ private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateA
                     continue;
                 }
 
-                if (DiagnosticsAreUnchanged(documentToPreviousDiagnosticParams, document))
-                {
-                    context.TraceInformation($"Diagnostics were unchanged for document: {document.FilePath}");
+                TReport report;
 
-                    // Nothing changed between the last request and this one.  Report a (null-diagnostics,
-                    // same-result-id) response to the client as that means they should just preserve the current
-                    // diagnostics they have for this file.
-                    var previousParams = documentToPreviousDiagnosticParams[document];
-                    progress.Report(CreateReport(previousParams.TextDocument, diagnostics: null, previousParams.PreviousResultId));
-                }
-                else
+                // Ensure the read to _documentIdToLastResultId to get the cached resultId and subsequent write
+                // to _documentIdToLastResultId after calculating new diagnostics happens in a single 'transaction'
+                // to prevent us from overwriting data in _documentIdToLastResultId with stale results.
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
The particular issue I noticed is an issue where _documentIdToLastResultId is cleared for a document in between the read in DiagnosticsAreUnchanged and a write to it in ComputeAndReportCurrentDiagnosticsAsync for a single LSP pull request.

Consider the scenario
1.  A change in the editor triggers the OnDiagnosticsUpdated event to clear resultId for File.cs
2.  LSP sends a document pull for File.cs, we have no cached diagnostics (cleared by 1) so we _begin_ calculation.
3.  A scenario like stop debugging causes OnDiagnosticsUpdated to be triggered and resultId is cleared _before_ the calculation in 2) completes.  
4.  The calculation in 2) completes and updates the resultId for File.cs, _overwriting_ the clear that happened in 3) with stale results.
5.  The client gets the result from 4).  Now the client sends the resultId from 4 for the next request, and we happily return them the stale cached results _for all subsequent requests_ until we get a new OnDiagnosticsUpdated (like when the user types).

With my fix, OnDiagnosticsUpdated will always clear the cached resultId _after_ or _before_ a request has entirely completed.  
1.  If the LSP pull request for a change comes in _after_ the OnDiagnosticsUpdated for the same change, then it sees the cleared resultId and re-calculates.  
2.  If the LSP pull request for a change comes in _before_ the OnDiagnosticsUpdated for the same change, then the pull request will return the stale cached diagnostics, but the subsequent polling by the client will soon query again at which point the stale diagnostics will have been cleared.